### PR TITLE
feat[lang]: use keyword arguments for struct instantiation

### PR DESCRIPTION
**Pull request by tserg**
### What I did

closes: https://github.com/vyperlang/vyper/issues/2381

Change the instantiation of a struct from a dictionary to a callable with kwargs.

### How I did it

Update the analysis and codegen, as well as grammar.

### How to verify it

See tests

### Commit message

```
feat: use keyword arguments for struct instantiation

This commit changes struct instantiation from taking a single dict
argument to taking keyword arguments.

For backwards compatibility (and to ease users into v0.4.0), the old
syntax is still supported, but a warning will be emitted.
```

### Description for the changelog

Use kwargs for struct instantiation.

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

**Review by fubuloubu — APPROVED** [review]
love this change

**Comment by fubuloubu** [comment]
btw isn't this tracked in a VIP somewhere? please link

**Comment by tserg** [comment]
> btw isn't this tracked in a VIP somewhere? please link

Yes, I have updated the top-level post with the link to the VIP.

**Comment by fubuloubu** [comment]
> > btw isn't this tracked in a VIP somewhere? please link
> 
> Yes, I have updated the top-level post with the link to the VIP.

if you use `closes: #X` it will link and close the VIP as completed when this is merged

**Comment by tserg** [comment]
> > > btw isn't this tracked in a VIP somewhere? please link
> > 
> > 
> > Yes, I have updated the top-level post with the link to the VIP.
> 
> if you use `closes: #X` it will link and close the VIP as completed when this is merged

I was under the impression that this PR partially addressed the issue, as there was a separate discussion about having default values, but then I noticed your last [comment](https://github.com/vyperlang/vyper/issues/2381#issuecomment-882801070). Thanks anyway!

**Review by charles-cooper — CHANGES_REQUESTED** [review]
this is looking good. we should add a warning if somebody tries to use the old way, instead of failing. i don't think we should break compatibility for at least a few releases, so people have time to ease into it.

i think we can keep support for the old way by adding an AST transformation, when `Call`s are constructed you can check if the argument is a dict, and transform it.

File: vyper/ast/parse.py [hunk 1]
```diff
@@ -341,6 +341,25 @@ def visit_Expr(self, node):
...
 
+    def visit_Call(self, node):
+        """
+        Convert structs declared as `Dict` node for vyper < 0.4.0 to kwargs.
+        """
+        if len(node.args) == 1 and isinstance(node.args[0], python_ast.Dict):
+            dict_ = node.args[0]
+            kw_list = []
+
+            for key, value in zip(dict_.keys, dict_.values):
```
Comment (charles-cooper) [inline vyper/ast/parse.py:352]: even though we are guaranteed by the parser that `.keys` and `.values` have the same length, i'd like to have a sanity check that `assert len(dict_.keys) == len(dict_.values)`

Comment (tserg) [reply vyper/ast/parse.py:360]: done!

File: tests/functional/syntax/test_structs.py [hunk 2]
```diff
@@ -572,7 +555,16 @@ def get_y() -> int128:
 
 @external
 def foo():
-    bar: C = C({a: 1, b: block.timestamp})
+    bar: C = C(a=1, b=block.timestamp)
+    """,
+    # backwards compatibility for vyper <0.4.0
```
Comment (charles-cooper) [inline tests/functional/syntax/test_structs.py:560]: maybe we should add a check that the warning is emitted

File: vyper/semantics/types/user.py [hunk 3]
```diff
@@ -381,38 +381,37 @@ def to_abi_arg(self, name: str = "") -> dict:
         components = [t.to_abi_arg(name=k) for k, t in self.member_types.items()]
         return {"name": name, "type": "tuple", "components": components}
 
-    # TODO breaking change: use kwargs instead of dict
-    # when using the type itself (not an instance) in the call position
-    # maybe rename to _ctor_call_return
     def _ctor_call_return(self, node: vy_ast.Call) -> "StructT":
-        validate_call_args(node, 1)
-        if not isinstance(node.args[0], vy_ast.Dict):
+        if len(node.args) > 0:
```
Comment (charles-cooper) [inline vyper/semantics/types/user.py:385]: is there a test for this?

Comment (charles-cooper) [reply vyper/semantics/types/user.py:385]: I don't think it's unreachable though, e.g.`MyStruct(a, x=y)`

Comment (tserg) [reply vyper/semantics/types/user.py:385]: oh ic, I have just added a syntax test.

**Review by charles-cooper — APPROVED** [review]

**Comment by pcaversaccio** [comment]
I'm not sure whether this is intended behaviour or simply a regression tbh but if I use now the same variable name for the kwarg, it will fail:

image

```vy
struct Batch:
    target: address
    allow_failure: bool
    call_data: Bytes[max_value(uint16)]


struct Result:
    success: bool
    return_data: Bytes[max_value(uint8)]


@external
def multicall(data: DynArray[Batch, max_value(uint8)]) -> DynArray[Result, max_value(uint8)]:
    results: DynArray[Result, max_value(uint8)] = []
    return_data: Bytes[max_value(uint8)] = b""
    success: bool = empty(bool)
    for batch: Batch in data:
        if (batch.allow_failure == False):
            return_data = raw_call(batch.target, batch.call_data, max_outsize=255)
            success = True
            results.append(Result({success: success, return_data: return_data})) # This will pass
            results.append(Result({success=success, return_data=return_data})) # This will raise
        else:
            success, return_data = \
                raw_call(batch.target, batch.call_data, max_outsize=255, revert_on_failure=False)
            results.append(Result({success: success, return_data: return_data})) # This will pass
            results.append(Result({success=success, return_data=return_data})) # This will raise
    return results
```

Since the intention was to make it backward-compatible, using e.g. `success=success` should also compile IMO.

**Comment by tserg** [comment]
```
results.append(Result({success=success, return_data=return_data})) # This will raise
```
The curly brackets are dropped in the new syntax, so this should compile:
```
results.append(Result(success=success, return_data=return_data))
```

**Comment by pcaversaccio** [comment]
> ```
> results.append(Result({success=success, return_data=return_data})) # This will raise
> ```
> 
> The curly brackets are dropped in the new syntax, so this should compile:
> 
> ```
> results.append(Result(success=success, return_data=return_data))
> ```

Ah interesting, that indeed compiles, thx! Added this new feature to snekmate modules branch: https://github.com/pcaversaccio/snekmate/pull/207/commits/0c3a20308ef46cfe51cef4912704faa608bea712. Vey nitpick: we should always follow [PEP-8](https://peps.python.org/pep-0008/#other-recommendations) where it states:
> Don’t use spaces around the `=` sign when used to indicate a keyword argument

The two following tests don't follow this:

- https://github.com/vyperlang/vyper/pull/3777/files#diff-cb5228f23a8a9317e7de9bc3780d62cdfd69bb2655261be60e55143a874615f4
- https://github.com/vyperlang/vyper/pull/3777/files#diff-51af4b3b600a938c59a20303d4a83f82f12746942d85c99bc7284bad5abfdc43

**Comment by pcaversaccio** [comment]
Also, I think it would make sense to update the docs [here](https://docs.vyperlang.org/en/latest/types.html#structs) with the new syntax.

**Comment by tserg** [comment]
> Also, I think it would make sense to update the docs [here](https://docs.vyperlang.org/en/latest/types.html#structs) with the new syntax.

Good point, I will follow up on this and the tests. Thanks!

**Pull request merged**